File: 03_Stack/carFleet.js
// Function to calculate the number of car fleets
var carFleet = function(target, position, speed) {
    // Get the coordinates representing time to reach the target for each car
    const coordinates = getCoordinates(target, position, speed);

    // Search for car fleets in descending order of time to reach the target
    return searchDescending(coordinates);
};

// Function to get coordinates representing time to reach the target for each car
var getCoordinates = (target, position, speed) => position
    .map((_position, index) => [ _position, speed[index] ])        // Create pairs of position and speed
    .sort(([ aPosition ], [ bPosition ]) => bPosition - aPosition) // Sort the pairs in descending order based on position
    .map(([ _position, _speed ]) => (target - _position) / _speed); // Calculate time to reach the target for each car

// Function to search for car fleets in descending order of time to reach the target
var searchDescending = (coordinates, previous = 0, fleets = 0) => {
    // Iterate through the coordinates in descending order
    for (const coordinate of coordinates) {
        // Check if the current coordinate is greater than the previous one
        const isPreviousLess = previous < coordinate;
        if (!isPreviousLess) continue; // Skip to the next iteration if the previous coordinate is greater or equal

        // Update the previous coordinate and increment the fleet count
        previous = coordinate;
        fleets++;
    }

    // Return the total number of car fleets
    return fleets;
}

// implementation
console.log("Leetcode 853: Car Fleet\n");

// Example 1
var target1 = 12;
var position1 = [10, 8, 0, 5, 3];
const speed1 = [2, 4, 1, 1, 3];
console.log(`Example 1: target = ${target1}, position = [${position1}], speed = [${speed1}]`);
console.log("Output:", carFleet(target1, position1, speed1));

// Example 2
var target2 = 10;
const position2 = [3];
const speed2 = [3];
console.log(`Example 1: target = ${target2}, position = [${position2}], speed = [${speed2}]`);
console.log("Output:", carFleet(target2, position2, speed2));

// Example 3
var target3 = 100;
const position3 = [0, 2, 4];
const speed3 = [4, 2, 1];
console.log(`Example 1: target = ${target3}, position = [${position3}], speed = [${speed3}]`);
console.log("Output:", carFleet(target3, position3, speed3));

File: 03_Stack/car_fleet.py
class Solution:
    def carFleet(self, target: int, position: list[int], speed: list[int]) -> int:
        # Create pairs of position and speed and sort them in reverse order based on position
        pair = [(p, s) for p, s in zip(position, speed)]
        pair.sort(reverse=True)

        # Initialize a stack to keep track of time to reach the target for each car
        stack = []

        # Iterate through the sorted pairs in reverse order
        for p, s in pair:
            # Calculate the time to reach the target for the current car
            stack.append((target - p) / s)

            # Check if there are at least two cars in the stack and the current car is slower than the previous car
            if len(stack) >= 2 and stack[-1] <= stack[-2]:
                stack.pop()  # Remove the previous car from the stack as it is fleet with the current car

        # The length of the stack represents the number of car fleets
        return len(stack)
    

# implementation 
answer = Solution()

print("Leetdcode 853: Car Fleet\n")

# Example 1
target_1 = 12
position_1 = [10, 8, 0, 5, 3]
speed_1 = [2, 4, 1, 1, 3]
print(f"Example 1: target = {target_1}, position = {position_1}, speed = {speed_1}")
print("Output:", answer.carFleet(target_1, position_1, speed_1))

# Example 2
target_2 = 10
position_2 = [3]
speed_2 = [3]
print(f"Example 2: target = {target_2}, position = {position_2}, speed = {speed_2}")
print("Output:", answer.carFleet(target_2, position_2, speed_2))


# Example 3
target_3 = 100
position_3 = [0, 2, 4]
speed_3 = [4, 2, 1]
print(f"Example 3: target = {target_3}, position = {position_3}, speed = {speed_3}")
print("Output:", answer.carFleet(target_3, position_3, speed_3))
